timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/AsyncLockService.java

```diff
@@ -89,8 +92,14 @@ private void scheduleExpiredLockReaper() {
     }
 
     public AsyncResult<Void> waitForLocks(UUID requestId, Set<LockDescriptor> lockDescriptors, TimeLimit timeout) {
-        OrderedLocks orderedLocks = locks.getAll(lockDescriptors);
-        return lockAcquirer.waitForLocks(requestId, orderedLocks, timeout);
+        AsyncResult<Void> result = waitRequests.computeIfAbsent(requestId, ignored -> {
+            OrderedLocks orderedLocks = locks.getAll(lockDescriptors);
+            return lockAcquirer.waitForLocks(requestId, orderedLocks, timeout);
+        });
+        result.onComplete(() -> {
+            waitRequests.remove(requestId);
+        });
+        return result;
```

Comment:
After switching to use the AwaitedLocksCollection, this should read something like this?

```java
return awaitedLocks.getExistingOrAwait(requestId, () -> {
            OrderedLocks orderedLocks = locks.getAll(lockDescriptors);
            return lockAcquirer.waitForLocks(requestId, orderedLocks, timeout);
        });
```